[1단계 - 자동차 경주 구현] 호호(유현호) 미션 제출합니다.

yhh1056 opened the pull request:
안녕하세요 카일! 호호입니다. 첫 리뷰 요청이라 떨리네요.
문자열 계산기에서 작은 단위로 TDD를 하는 방식을 배웠지만 자동차 경주 미션에 적용하는 게 쉽지 않았어요.

프리코스 과정에서 객체에게 메시지를 보내라는 피드백을 받아서 이번 자동차 경주 미션에도 적용해 보려고 했습니다.
출력을 위한 get 메서드가 아니라면 최대한 지양하고 객체가 스스로 판단할 수 있도록 연습해 보았어요.

질문 하나 하겠습니다!

처음 구현을 시작할 때 도메인으로 자동차가 떠올랐습니다. '자동차가 주어진 조건에 따라 전진할 수 있다'라는 기능만 생각했어요.
구현하다 보니 자동차는 스스로 위치(position)를 비교할 수 있어야 하고, 위치가 동일한지 판단할 수 있어야 되더라고요. 기능 목록을 작성할 때 미리 떠올랐다면 더 좋았을 텐데 말이죠. 기능 요구사항을 보고 객체의 책임과 역할을 잘 떠올리려면 어떤 연습이 도움이 될까요?

Commit:
feat(study): JUnit 학습

Commit:
feat(calculator): 문자열 덧셈 계산기

Commit:
refactor(calculator): 메소드 분리 및 네이밍 수정

Commit:
docs(readme): 기능구현 목록 작성

Commit:
feat(racingcar): 이름으로 자동차 생성

Commit:
feat(advance): 조건에 따른 자동차 전진 기능 추가

Commit:
feat(random): 0에서 9사이의 랜덤한 숫자를 생성

Commit:
feat(name): 자동차의  이름을 쉼표로 구분

Commit:
feat(name): 이름 앞 뒤 공백 제거

Commit:
feat(name): 공백 제거 테스트 클래스 이동

Commit:
feat(name): 자동차 이름이 5자를 넘을 경우 IAE발생

Commit:
refactor(racingcar) cars에서 자동차들을 생성하도록 수정

Commit:
refactor(input): 같은 이름이 존재하는지 검증

Commit:
feat(input): 자동차의 이름을 입력받아 생성

Commit:
feat(input): 경주 횟수 입력

Commit:
feat(input): 입력된 경주 횟수 검증

Commit:
feat(racingcar): 시도 횟수만큼 자동차 경주

Commit:
feat(output): 진행 결과를 출력

Commit:
feat(winner): 가장 많이 전진한 자동차 판별

Commit:
feat(winner): 여려명일 경우에도 우승자 출력

Commit:
refactor(packages): 패키지 분리

Commit:
refactor(car): 자동차의 위치를 반환하는 메소드 구현

Commit:
refactor(racinggame): 전진 시도 메소드 분리, Runnable 제거

Commit:
fix(winner): 가장 많이 전진한 자동차가 우승하도록 수정

기존에 포지션이 내림차순으로 정렬되어 가장 적게 전진한 자동차가 전진되어 오름차순으로 정렬하도록 수정

Commit:
refactor(name): 이름 생성 메소드명 수정

of 보다 더 의미있는 네이밍으로 생성하도록 변경하였습니다.

Commit:
refactor(trial): 시도 횟수를 값 객체로 수정

Commit:
refactor(winner): 우승자를 값 객체로 수정

Commit:
refactor(racinggame): 게임 과정, 결과 출력

Commit:
feat(car): 자동차가 전진 숫자에 대한 유효성을 판단

Commit:
refactor(name): 인풋을 names로 변환해주는 메소드 이름 변경

Commit:
refactor(name): 우승자가 존재하는지 테스트할 수 있도록 변경

Commit:
refactor(cars): 조회된 자동차의 상태 값이 변경되지 않도록 CarValue 생성

Commit:
refactor(package): 데이터를 전달하는 객체를 dto 패키지로 분리

Commit:
refactor(package): 데이터를 전달하는 객체를 dto 패키지로 분리

KIMSIYOUNG commented on README.md:
```diff
@@ -5,3 +5,22 @@
 
 - [온라인 코드 리뷰 과정](https://github.com/woowacourse/woowacourse-docs/blob/master/maincourse/README.md)
+
+## 기능 구현 목록
```
구체적인 요구사항 작성을 잘 해주셨네요. 👏

계산기도 코드에 포함되어 있으니 계산기 요구사항도 작성해주면 좋을 것 같아요.

추가로 개인적인 생각인데, 각 기능을 쭉 보고 어떤 프로그램이구나 라는걸 알 수 있지만 프로그램이 기본적으로 어떤 것이다 라는 정의도 함께 있으면 처음 보더라도 이해가 쉬울 것 같네요🙂

yhh1056 replied:
친절한 리드미가 되도록 수정할게요!

KIMSIYOUNG replied:
실행결과가 같이 있으니까 어떤 프로그램인지 정말 잘 보이는 것 같아요👏

KIMSIYOUNG commented on src/main/java/calcaulator/StringCalculator.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    private static int stringToInt(String string) {
+        return Integer.parseInt(string);
```
숫자를 파싱하는 과정에서 발생하는 케이스가 누락된 것 같아요. 

로직 + 테스트도 같이 추가해주세요🙂

yhh1056 replied:
확인 감사합니다! 테스트와 같이 추가할게요🙂

KIMSIYOUNG commented on src/main/java/calcaulator/StringCalculator.java:
```diff
@@ -0,0 +1,66 @@
+            return DEFAULT_VALUE_FOR_NULL_OR_EMPTY;
+        }
+
+        if (hasSingleNumber(text)) {
```
해당 메소드는 길이가 하나임을 구현하고 있지만 이름은 하나의 수를 가지고 있다라고 네이밍되어 있어요.

조금 더 메소드의 역할을 드러낼 수 있게 변경하면 좋을 것 같아요.

yhh1056 replied:
코드를 다시 보면서 메소드의 이름이 적절한지 확인해봐야겠어요. 좀 더 의미있는 이름으로 수정할게요! 👍

KIMSIYOUNG commented on src/main/java/calcaulator/StringCalculator.java:
```diff
@@ -0,0 +1,66 @@
+
+    private static final int MINIMUM_NUMBER = 0;
+    private static final int MINIMUM_LENGTH = 1;
+    private static final int DEFAULT_VALUE_FOR_NULL_OR_EMPTY = 0;
```
동일한 0이지만 의미가 다르기에 다른 상수로 분리해주신 부분이 매우 좋은 것 같습니다!

KIMSIYOUNG commented on src/main/java/racingcar/vo/Name.java:
```diff
@@ -0,0 +1,51 @@
+
+import java.util.Objects;
+
+public class Name {
```
꼼꼼한 역할 및 객체분리 👏

KIMSIYOUNG commented on src/main/java/racingcar/vo/Name.java:
```diff
@@ -0,0 +1,51 @@
+        return name.trim();
+    }
+
+    public static Name create(String name) {
```
정적팩토리메소드를 사용해주셨네요. 

혹시 어떤 의미를 위해 create라는 것을 사용해주셨을까요?

추가로 생성자가 있음에도 불구하고 정적팩토리메소드의 형태를 사용하는 경우도 같이 찾아보면 좋을 것 같아요.

yhh1056 replied:
처음엔 관례적으로 사용하는 of라는 이름을 사용했어요. 좀 더 분명한 이름을 지어주는게 가독성에 좋지 않을까 생각이 들어 String 타입으로  Name을 생성하므로 create라는 이름으로 변경했어요. 다른 네이밍으로는 어떤게 있을까요?

정적팩토리메소드의 형태를 사용하는 경우를 찾아보았는데요. 예를 들면 다음과 같은 예제로 사용할 수 있을 거 같아요!

```java
public static Car hohoCar() {
    return new Car("hoho");
}
```

KIMSIYOUNG replied:
앗 제가 의미를 조금 잘못 전달드린 것 같아요. 

저는 `생성자가 있음에도 불구하고` 이를 사용하지 않고 정적메소드를 사용한 이유가 궁금했어요.

yhh1056 replied:
정적 메소드를 사용하면 이름을 가지고, 캡슐화를 할 수 있고, 생성시 최적화를 할 수 있는 장점이 있지만 
지금 제 코드에서 정적메소드를 사용하여 생성하는 것과 생성자를 사용하는 것이 큰 기능적 차이가 없어요. 트레이드 오프였는데 저는 정적 메소드로 생성하는 방식이 거부감이 들지 않고 읽기 편해서 사용했어요.

KIMSIYOUNG commented on src/main/java/racingcar/vo/Name.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    private String removeWhiteSpaces(String name) {
+        return name.trim();
```
트림은 양쪽의 공백을 지우는 것인데, 메소드명은 전체 공백을 지우는 것처럼 되어있어요. 

의미를 조금 더 명확히 하면 좋을 것 같아요.

yhh1056 replied:
여기도 명확한 이름으로 변경할게요! 👍

KIMSIYOUNG replied:
각 메소드의 기능을 세분화하기 위해서 작은 일이더라도, 메소드로 분리한 부분이 인상적인 것 같아요👏

현재 미션에서는 잘 수행해주신 것 같고, 이 기능이 메소드로 분리되어야 하는가에 대해서도 호호만의 의견이 있으면 좋을 것 같아요. 

`trim`이라는 것만으로도 충분히 의미를 잘 드러내고 있지만, 이름을 트림하는건 우리의 어플리케이션내에서 다른 의미를 포함하고 있다면 그 의미가 드러나도록 메소드를 분리하는 것도 좋을 것이에요. 다만 단순한 트림이라면 메소드를 분리하는게 맞을지 고민해보면 좋을 것 같습니다🙂

yhh1056 replied:
'우리의 어플리케이션내에서 다른 의미를 포함하고 있다면 그 의미가 드러나도록 메소드를 분리하는 것도 좋을 것이에요.' 이 부분이 정말 공감이 가네요! 

제 코드에서는 단순히 trim의 기능을 하고있어서 분리하지 않고 가는게 제 생각이 될 거 같아요. 
다만 다음에 어플레케이션만의 다른 기능이거나 다른 의미로 사용된다면 분리를 시도해야겠어요. 감사합니다 👍👍👍

KIMSIYOUNG commented on src/test/java/racingcar/domain/CarTest.java:
```diff
@@ -0,0 +1,93 @@
+        Car car = new Car(Name.create("test"));
+
+        // when
+        car.advance(number);
```
테스트하기 어려운 값을 파라미터로 올려 유의미한 테스트를 작성해주셨네요🙂

랜덤을 테스트할 수 있는 다른 방법은 어떤게 있을지 찾아보면 좋을 것 같아요.

yhh1056 replied:
랜덤을 테스트하는 방법이 까다롭더라구요. 다른 방법도 고민해보겠습니다!😃

KIMSIYOUNG replied:
넵넵! `랜덤을 테스트한다`라는 말의 의미가 정말 랜덤한 것이 랜덤하게 동작하는지를 테스트하는 것인지, 아니면 랜덤한 것은 우리가 제어할 수 없는 것이니 그 부분은 제외하고 그 부분에 의존하고 있는 것들이 올바르게 동작하는지를 테스트하는 것인지 고민해보면 좋을 것 같아요. `결국 우리가 테스트해야하는 것은 무엇인가?` 의 질문의 답이 될 것 같아요.

잘 변경해주신 것 같아요🙂 그래도 아직 혼선이 조금 있는 것 같아서, 위의 질문에 대해서 답을 내린다면 호호의 코드에 스스로 조금 더 확신을 가질 수 있을 것 같네요.

KIMSIYOUNG commented on src/main/java/racingcar/vo/Name.java:
```diff
@@ -0,0 +1,51 @@
+
+    private final String name;
+
+    public Name(String name) {
```
VO와 도메인 DTO를 모두 나눠주셨는데요.

호호는 어떤 기준으로 각각을 분리하셨는지 조금 더 설명해주실 수 있을까요?

yhh1056 replied:
도메인은 자동차 경주 미션에서 핵심 로직을 담당하는 객체들을 도메인으로 분리하였는데요. 도메인이 가장 어려운 거 같아요. 전진하거나 멈추면서 이동하는 `Car`클래스와 자동차들을 움직이게 하고 우승자를 판별해 주는 `Cars`클래스를 도메인으로 생각했어요.

한번 생성되면 변경이 불가능하고, 객체가 가지는 상태 값이 서로 동일하다면 같은 객체로 판단할 수 있는 객체를 VO로 분리하였어요. 이번 미션에서 `Name`과 `Trial`은 이름이 서로 동일하거나 시도 횟수가 동일하면 같은 객체로 판단할 수 있기 때문에 VO로 분류했어요.

DTO는 뜻대로 데이터를 전달하기 위해 사용되는 객체를 DTO로 분류했어요. 사용자의 입력값을 전달받거나 결괏값을 출력하기 위해 전달되는 객체를 DTO로 사용하였어요. 자동차 경주를 진행할 때 각 라운드마다 자동차를 전달해서 전진한 상태를 출력하는데 `Car` 객체를 그대로 넘기니까 `Output`에서도 `Car`의 상태를 변경할 수 있더라고요. 그래서 DTO를 사용해서 전달했어요.

KIMSIYOUNG replied:
좋은 기준으로 분리하신 것 같아요. `네임`이라는 것에 해당 질문을 드렸던 이유는 VO와 도메인의 경계는 어디일까? 에 대해서 고민해보면 좋을 것 같아서 말씀드렸어요🙂

VO는 도메인이 될 수 있는가? 도메인은 VO가 될 수 있는가? 된다면 어떤 상관관계를 갖는게 이상적일지까지에 대해서도 한 번 고민해보시면 좋을 것 같아요.

yhh1056 replied:
정말 좋은 고민인 것 같아요! 한 번 생각해보고 말씀드릴게요!

yhh1056 replied:
도메인은 어플레케이션의 비즈니스 행위를 가지고 책임을 수행한다고 생각해요. VO가 도메인이 되거나 도메인이 VO가 되는 건 서로가 가진 특징을 만족하게 된다면 도메인이 VO가 될 수 있고 VO도 도메인이 된다고 생각해요.

하지만 VO는 도메인의 상태를 표현해 주는 객체인데, 상태 자체만으로 비즈니스의 행위를 할 수 없다고 생각해요. 미션에서 자동차 경주의 우승자를 VO로 만들었는데요. 비즈니스 요구사항이 우승자를 출력하기 때문에 단순히 우승자라는 값으로만 사용되었어요. 불변을 보장하죠. 

추후에 요구사항으로 우승자에게 트로피를 주거나 상금을 주는 요구사항이 추가된다면 우승자는 시상을 담당하는 객체와 협력을 해야 되기 때문에 비즈니스 행위를 가지게 되고 상태가 변경될 수 있기 때문에 VO라는 특징을 잃게 되고 도메인이라 부를 수 있어 보여요.

객체(도메인 모델)들이 협력을 하기 위해 필요한 값 또는 객체의 상태를 표현하기 위한 값으로 VO를 사용한다면 도메인과 VO가 적절한 상관관계를 가지고 있다고 생각해 보았습니다. 카일은 어떻게 생각하는지 듣고 싶습니다!

KIMSIYOUNG reviewed (changes requested):
안녕하세요 호호
전체적으로 객체 / 메소드 모두 잘 분리해주신 것 같아요👏
간단한 리뷰 남겼으니 확인후에 반영해주세요.

`기능 요구사항을 보고 객체의 책임과 역할을 잘 떠올리려면 어떤 연습이 도움이 될까요?`

- 이 부분에 대해서는 사실 언제나 쉽지 않은 것 같아요. 어떤 프로그램을 완성하기 위해 필요한 요구사항들을 정리하고 객체레벨에서의 고민이전에, 어떤 역할들이 필요한지 고민해보는 것도 좋은 방법일 것 같아요. 하지만 이렇게 고민을 충분히 했다고 하더라도 구현하다보면 내가 몰랐던 엣지 케이스가 발생하고, 도메인에 대한 이해도가 더 높아지기 때문에 더 나은 설계법이 생각날 확률이 높아요. 그러다보니 저는 첫 구현에 최대한 완벽하게 하겠다라는 마음보단 구현하고 리팩토링하고 리팩토링하고 하는 것을 반복하면서 개선해나가는게 잘 맞는 것 같아요. 
- 구현 이전에 다이어그램이나 구조화를 해서 요구사항을 파악할 때 저는 도움이 많이 되었던 것 같아요. 다만 개발을 해보기전에 고민한 것과 실제 개발하면서 알게되는 도메인 지식은 분명 있을 수 밖에 없기 때문에 구현 - 리팩토링 - 구현 -리팩토링 사이클에 익숙해지는 것도 중요한 것 같아요.

Commit:
docs(readme): 문자열 계산기, 자동차 경주게임 설명 추가

Commit:
feat(calculator): 문자열이 숫자인지 검증하도록 추가

Commit:
refactor(naming): 의미를 표현할 수 있도록 네이밍 수정

Commit:
refactor(cars): 자동차들의 전진 관심 분리

Cars 객체는 주어지는 숫자에 따라 자동차들을 전진시키는 역할을 수행하고 RacingGame 객체가 랜덤한 숫자를 생성하여 Cars객체에게 자동차를 움직이게 명령하도록 수정하였습니다.

자동차들이 스스로 움직이지 않고 외부에 주어지는 숫자로 움직이다보니 자동차의 움직임을 제어할 수 있어서 테스트를 진행할 수 있었습니다.

Commit:
refactor(number): 랜덤 숫자 네이밍 수정과 음수 테스트

yhh1056 commented:
안녕하세요!
피드백 주신 대로 역할을 들어낼 수 있는 네이밍에 대해 고민해 보고 수정했어요.

랜덤을 테스트하는 방법에 대해 고민해 보았는데요. 랜덤 객체에서 나오는 숫자를 반복문을 돌려서 테스트한다 해도 완전한 테스트라고 생각이 들진 않았어요. 그래서 랜덤으로 생성된 번호를 파라미터로 받는 곳에서 예외를 처리하도록 구현했어요.

[0c8d170](https://github.com/woowacourse/java-racingcar/pull/311/commits/0c8d1703b73479ea019b1ab9073bbce0aa6a6fab) : 자동차는 외부에서 랜덤 한 숫자를 받기 때문에 테스트하기가 용이했어요. 근데 매 라운드마다 자동차들이 이동하는 `move`메서드는 스스로 랜덤 숫자를 생성하고 이동시키기 때문에 외부에서 숫자들을 부여받는 식으로 변경해 보았어요. `Cars` 객체가 랜덤 숫자를 생성하고 움직이는 것보다 주어진 숫자에 맞게 자동차들을 이동시키다 보니 테스트를 할 때도 자동차들을 제어하기 수월했어요. 카일은 어떤 방법으로 랜덤을 테스트하셨나요?

KIMSIYOUNG commented on src/main/java/racingcar/io/Output.java:
```diff
@@ -21,7 +22,7 @@ private void printRacingProgress(CarValue car) {
         StringBuilder builder = new StringBuilder();
         for (int i = 0; i < position; i++) {
-            builder.append("-");
+            builder.append(CAR_ADVANCE_EXPRESSION);
```
사소한 부분이지만 차의 이동을 어떤 형태로 표현할 것이다 라는 것을 상수화하니까 의미 파악이 용이하네요👏

KIMSIYOUNG commented on src/test/java/racingcar/domain/CarsTest.java:
```diff
@@ -1,30 +1,96 @@
 
 public class CarsTest {
+
+    final int advanceValue = 5;
```
이것 또한 상수로 분류하는게 좋을 것 같아요🙂

KIMSIYOUNG commented:
안녕하세요 호호!
리뷰 잘 반영해주셨네요🙂
질문으로 적어주신 내용은 코멘트에 추가했고, 조금 더 고민해보면 좋을만한 것들 같이 남겼어요!
요번 미션은 충분히 잘 하신 것 같아서 머지할게요! 추가적인 코멘트는 다음 미션에서 반영해주시고, 궁금하신 내용은 언제든 말씀주세요!

KIMSIYOUNG reviewed (commented):
안녕하세요 호호!
리뷰 잘 반영해주셨네요🙂
질문으로 적어주신 내용은 코멘트에 추가했고, 조금 더 고민해보면 좋을만한 것들 같이 남겼어요!
요번 미션은 충분히 잘 하신 것 같아서 머지할게요! 추가적인 코멘트는 다음 미션에서 반영해주시고, 궁금하신 내용은 언제든 말씀주세요!

yhh1056 reviewed (commented):
항상 코드에 이유가 있어야 한다고 하는데 이번 카일에게 리뷰를 받으면서 코드를 작성하면서 더 의식적으로 고민해보면서 코드를 작성해야겠다고 느꼈어요. 다음 미션에서 더 신경써봐야겠어요. 감사합니다 😃😃
